src/public_api/functions.rs

```diff
@@ -2,29 +2,45 @@ use rustc_span::def_id::DefId;
 
 use rustc_middle::ty::TyCtxt;
 
-use crate::diagnosis::{DiagnosisCollector, DiagnosticGenerator};
+use crate::{
+    comparator::Diff,
+    diagnosis::{DiagnosisCollector, DiagnosticGenerator},
+    glue::Change,
+};
 
-#[derive(Clone, Copy, Debug, PartialEq)]
-pub(crate) struct FnMetadata(pub(crate) DefId);
+use super::ApiItem;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct FnMetadata {
+    arg_num: usize,
+    path: String,
+}
 
 impl FnMetadata {
-    pub(crate) fn new(id: DefId) -> FnMetadata {
-        FnMetadata(id)
+    pub(crate) fn new(tcx: &TyCtxt, id: DefId) -> FnMetadata {
+        let arg_num = tcx.fn_sig(id).skip_binder().inputs_and_output.len();
+        let path = tcx.def_path(id).to_string_no_crate_verbose();
+
+        FnMetadata { arg_num, path }
+    }
+
+    pub(crate) fn path(&self) -> &str {
+        self.path.as_str()
     }
-}
 
-impl DiagnosticGenerator for FnMetadata {
-    fn def_id(&self) -> DefId {
-        self.0
+    pub(crate) fn generate_changes(prev: FnMetadata, next: FnMetadata) -> Option<Change> {
```

Comment:
we need a better name

side note: as this function looks very similar to what `ApiItem::generate_changes` does, is it a good idea to put them in a trait, to show that they expose the same set of properties?